src/genotyper.cc

```diff
@@ -347,14 +335,96 @@ static Status translate_genotypes(const genotyper_config& cfg, const unified_sit
     return Status::OK();
 }
 
+/// Given a set of records overlapping a site, we find the
+/// "representative record(s)" for a given site. The representative
+/// record(s) found are returned via the ans vector. This function also
+/// modifies 2 genotyper substrates: failed_ref_depth and genotypes.
+/// Failed_ref_depth tracks the ref-records for a given sample has failed
+/// the minimal depth requirements (if any).
+/// Modifications to genotypes vector is explained below
+/// The representative record(s) are expected to be passed into translate_genotypes for genotyping
+/// =====================================================
+/// Pre conditions:
+/// ans should be an empty vector when passed into the function, otherwise an invalid status will be raised
+/// ======================================================
+/// Expected behaviors:
+/// No records given
+///      Empty ans is returned, no modification to genotypes and failed_ref_depth vectors
+/// Records do not span the entire range of site
+///      Empty ans is returned, failed_ref_depth updated accordingly, no change to genotypes
+/// Records span entire range, and consist of all gvcf records
+///      ans contains the first gvcf record passed, failed_ref_depth updated accordingly, no change to genotypes
+/// Records span entire range, and consist of exactly 1 vcf variant record
+///      ans contains the 1 vcf variant record, failed_ref_depth updated accordingly, no change to genotypes
+/// Records span entire range, and consists of >1 vcf variant record
+///      Empty ans is returned, failed_ref_depth updated accordingly, genotypes for all samples in this dataset are updated to be lost calls with RNC=LostAllele
+
+Status find_rep_record(const genotyper_config& cfg, const unified_site& site,
+                       const vector<shared_ptr<bcf1_t>>& records, const string& dataset,
+                       const bcf_hdr_t* hdr, int bcf_nsamples,
+                       const map<int, int>& sample_mapping, vector<one_call>& genotypes,
+                       vector<bool>& failed_ref_depth, vector<shared_ptr<bcf1_t>>& ans) {
+
+
+    // The ans vector should be empty when passed in
+    if(!ans.empty()) {
+        return Status::Invalid("find_rep_record: A non-empty ans vector was given.");
+    }
+
+    Status s;
+    unique_ptr<AlleleDepthHelper> depth;
+
+    // To track the total range covered by the records
+    vector<range> record_rngs;
+    for (auto& record: records) {
+        record_rngs.push_back(range(record.get()));
+        S(AlleleDepthHelper::Open(cfg, dataset, hdr, record.get(), depth));
 
-Status genotype_site(const genotyper_config& cfg, MetadataCache& cache, BCFData& data,
-                     const unified_site& site,
+        for (int i=0; i<bcf_nsamples; i++) {
+            // TODO: Consider support for hom ref records
```

Comment:
It will be interesting to look at the residuals for the 8-samples to see if that case arises.